core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -1434,38 +1435,51 @@ public void executeSystemTask(WorkflowSystemTask systemTask, String taskId, long
 
             LOGGER.debug("Executing {}/{}-{}", task.getTaskType(), task.getTaskId(), task.getStatus());
             if (task.getStatus() == SCHEDULED || !systemTask.isAsyncComplete(task)) {
-                task.setPollCount(task.getPollCount() + 1);
+                task.incrementPollCount();
                 executionDAOFacade.updateTask(task);
             }
 
+            // load task data (input/output) from external storage, if necessary
             deciderService.populateTaskData(task);
 
-            // Stop polling for asyncComplete system tasks that are not in SCHEDULED state
-            if (systemTask.isAsyncComplete(task) && task.getStatus() != SCHEDULED) {
-                queueDAO.remove(QueueUtils.getQueueName(task), task.getTaskId());
-                return;
+            if (task.getStatus() == SCHEDULED) {
+                systemTask.start(workflow, task, this);
+            } else if (task.getStatus() == IN_PROGRESS) {
+                systemTask.execute(workflow, task, this);
             }
 
-            switch (task.getStatus()) {
-                case SCHEDULED:
-                    systemTask.start(workflow, task, this);
-                    break;
-
-                case IN_PROGRESS:
-                    systemTask.execute(workflow, task, this);
-                    break;
-                default:
-                    break;
+            if (task.getOutputData() != null && !task.getOutputData().isEmpty()) {
+                deciderService.externalizeTaskData(task);
             }
 
-            if (!task.getStatus().isTerminal()) {
+            // Update message in Task queue based on Task status
+            // Stop polling for asyncComplete system tasks that are not in SCHEDULED state
+            if (systemTask.isAsyncComplete(task) && task.getStatus() != SCHEDULED) {
```

Comment:
@apanicker-nflx Instead of removing the task in the next run, i changed the logic to remove it eagerly. Previously, an "async complete" task in `IN_PROGRESS` state is postponed. And in the next run, the "async completeness" is checked and removed from the queue.